src/pex-builder/builder/deploy.py

```diff
@@ -0,0 +1,267 @@
+) -> List[LocationBuild]:
+    location_builds = [
+        LocationBuild(
+            location=location,
+            deps_requirements=deps.get_deps_requirements(
+                location.directory, python_version=python_version
+            ),
+        )
+        for location in locations
+    ]
+
+    # dedup requirements so each is only built once
+    builds_for_requirements_hash: Dict[str, List[LocationBuild]] = {}
+    for location_build in location_builds:
+        requirements_hash = location_build.deps_requirements.hash
+        builds_for_requirements_hash.setdefault(requirements_hash, []).append(
+            location_build
+        )
+
+    # build each deps pex once and assign to all related builds
+    for requirements_hash in builds_for_requirements_hash:
+        builds = builds_for_requirements_hash[requirements_hash]
+        deps_requirements = builds[0].deps_requirements
+
+        # don't build deps.pex files that are already published
+        published_deps_pex = (
+            get_published_deps_pex_name(deps_requirements.hash) if upload_pex else None
+        )
+
+        if published_deps_pex:
+            logging.info(
+                "Found published deps.pex %r for requirements_hash %r, skipping rebuild.",
+                published_deps_pex,
+                deps_requirements.hash,
+            )
+
+            for location_build in builds:
+                location_build.published_deps_pex = published_deps_pex
+        else:
+            logging.info(
+                "No published deps.pex found for requirements_hash %r, will rebuild.",
+                deps_requirements.hash,
+            )
+            deps_pex_path = deps.build_deps_from_requirements(
+                deps_requirements, output_directory, python_version
+            )
+
+            for location_build in builds:
+                location_build.deps_pex_path = deps_pex_path
+
+    # build each source once
+    for location_build in location_builds:
+        location_build.source_pex_path = source.build_source_pex(
+            location_build.location.directory, output_directory, python_version
+        )
+
+    # compute pex tags
+    for location_build in location_builds:
+        deps_pex = (
+            location_build.deps_pex_path
+            if location_build.deps_pex_path
+            else location_build.published_deps_pex
+        )
+        if not deps_pex or not location_build.source_pex_path:
+            raise ValueError("No deps.pex or source.pex")
+
+        location_build.pex_tag = util.build_pex_tag(
+            [deps_pex, location_build.source_pex_path]
+        )
+    return location_builds
+
+
+def get_published_deps_pex_name(requirements_hash: str) -> Optional[str]:
+    return pex_registry.get_deps_pex_name_from_requirements_hash(requirements_hash)
+
+
+# TODO: Publish newer base image
+DEFAULT_BASE_IMAGE = "878483074102.dkr.ecr.us-west-2.amazonaws.com/dagster-cloud-agent-pre-license:pexdemo"
```

Comment:
So eventually this will be our actual `dagster-cloud-agent` image, right? Will we pin to latest?

Reply:
This is the code location server and runner image so I don't believe it's the same as the agent image. Users have no control over the serverless agent.

This image replaces the image that is currently built and uploaded from github actions and contains the user code. We will have a standard base image, or possibly a set of images (eg one image per python major.minor version).

Does this sound right @johannkm , @gibsondan?